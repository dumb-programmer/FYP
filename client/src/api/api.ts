export function login(data: { email: string; password: string }) {
  return fetch("http://localhost:3000/login", {
    method: "POST",
    body: JSON.stringify(data),
    headers: {
      "Content-Type": "application/json",
    },
    mode: "cors",
    credentials: "include",
  });
}

export function signup(data: {
  firstName: string;
  lastName: string;
  email: string;
  password: string;
}) {
  return fetch("http://localhost:3000/signup", {
    method: "POST",
    body: JSON.stringify(data),
    headers: {
      "Content-Type": "application/json",
    },
    mode: "cors",
  });
}

export function getMessages(chatId: string, page: number) {
  return fetch(`http://localhost:3000/chats/${chatId}/messages?page=${page}`, {
    mode: "cors",
    credentials: "include",
  });
}

export function searchMessages(chatId: string, query: string, page: number) {
  return fetch(
    `http://localhost:3000/chats/${chatId}/messages?query=${query}&page=${page}`,
    {
      mode: "cors",
      credentials: "include",
    }
  );
}

export function deleteMessage(chatId: string, messageId: string) {
  return fetch(`http://localhost:3000/chats/${chatId}/messages/${messageId}`, {
    method: "DELETE",
    mode: "cors",
    credentials: "include",
  });
}

export function sendPrompt(data: { prompt: string }, chatId?: string) {
  return fetch(`http://localhost:3000/chats/${chatId}`, {
    method: "POST",
    mode: "cors",
    credentials: "include",
    body: JSON.stringify(data),
    headers: {
      "Content-Type": "application/json",
    },
  });
}

export function getChats(page: number) {
  return fetch(`http://localhost:3000/chats?page=${page}`, {
    mode: "cors",
    credentials: "include",
  });
}

export function updateChat(chatId: string, name: string) {
  return fetch(`http://localhost:3000/chats/${chatId}`, {
    method: "PATCH",
    body: JSON.stringify({ name }),
    headers: {
      "Content-Type": "application/json",
    },
    credentials: "include",
    mode: "cors",
  });
}

export function deleteChat(chatId: string) {
  return fetch(`http://localhost:3000/chats/${chatId}`, {
    method: "DELETE",
    credentials: "include",
    mode: "cors",
  });
}
