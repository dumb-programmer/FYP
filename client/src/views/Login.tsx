import { useForm } from "react-hook-form";
import { zodResolver } from "@hookform/resolvers/zod";
import { LoginSchema } from "@/types/schema";
import { Link, useNavigate } from "react-router-dom";
import LoadingIcon from "@/components/LoadingIcon";
import AuthProviderBtns from "@/components/AuthProviderBtns";
import ErrorMessage from "@/components/ErrorMessage";
import { login } from "@/api/api";

export default function Login() {
    const { register, handleSubmit, setError, formState: { errors, isSubmitting } } = useForm({
        defaultValues: {
            email: "",
            password: "",
        },
        resolver: zodResolver(LoginSchema)
    });
    const navigate = useNavigate();

    return <main className="min-h-screen w-screen flex justify-center items-center">
        <div className="card p-10 shadow-md">
            <form className="flex flex-col gap-4 md:min-w-96" onSubmit={handleSubmit(async (data) => {
                const response = await login(data);
                if (response.ok) {
                    navigate("/");
                }
                else {
                    setError("root", {
                        message: "Invalid credentials"
                    });
                }
            })}>
                <h1 className="card-title text-2xl">Login</h1>
                <div className="form-control">
                    <label className="label" htmlFor="email">Email</label>
                    <input className="input input-bordered" id="email" {...register("email")} />
                    <ErrorMessage message={errors.email?.message} />
                </div>
                <div className="form-control">
                    <label className="label" htmlFor="password">Password</label>
                    <input className="input input-bordered" id="password" type="password" {...register("password")} />
                    <ErrorMessage message={errors.password?.message} />
                </div>
                <ErrorMessage message={errors.root?.message} />
                <button className="btn btn-primary">
                    {
                        isSubmitting ? <LoadingIcon size={25} stroke="white" /> : "Submit"
                    }
                </button>
                <p className="text-gray-500">Don't have an account? <Link className="link" to="/signup">Signup</Link></p>
                <div className="divider">OR</div>
                <AuthProviderBtns />
            </form>
        </div>
    </main>;
}