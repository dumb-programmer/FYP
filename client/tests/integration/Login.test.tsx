import React from "react";
import { render, screen } from "@testing-library/react";
import { expect, describe, it, vi } from "vitest";
import Login from "../../src/views/Login";
import { MemoryRouter, Route, Routes, useNavigate } from "react-router-dom";
import userEvent from "@testing-library/user-event";
import { login } from "../../src/api/api";


vi.mock("../../src/api/api", async () => ({ ...await vi.importActual("../../src/api/api"), login: vi.fn(() => ({ ok: true })) }))

vi.mock("react-router-dom", async () => ({ ...await vi.importActual("react-router-dom"), useNavigate: vi.fn() }))

const renderComponent = (redirect?: string) => render(
    <MemoryRouter initialEntries={[`/login${redirect ? `?redirect=${redirect}` : ""}`]}>
        <Routes>
            <Route path="/login" element={<Login />} />
        </Routes>
    </MemoryRouter>
);

describe("Login", () => {
    it("renders correctly", () => {
        const { container } = renderComponent();

        expect(container).toMatchSnapshot();
    });

    it("email input works", async () => {
        renderComponent();
        const input = screen.getByLabelText(/^email$/i);

        await userEvent.type(input, "test");

        expect(input.value).toEqual("test");
    });

    it("password input works", async () => {
        renderComponent();
        const input = screen.getByLabelText(/^password$/i);

        await userEvent.type(input, "test");

        expect(input.value).toEqual("test");
    });

    it("email input is validated", async () => {
        renderComponent();
        const input = screen.getByLabelText(/^email$/i);

        await userEvent.click(screen.getByRole("button"));

        expect(screen.getByText(/^email is required/i)).toBeInTheDocument();

        await userEvent.type(input, "test");

        await userEvent.click(screen.getByRole("button"));

        expect(screen.getByText(/^invalid email/i)).toBeInTheDocument();
    });

    it("password input is validated", async () => {
        renderComponent();
        const input = screen.getByLabelText(/^password$/i);

        await userEvent.type(input, "test");

        await userEvent.click(screen.getByRole("button"));

        expect(screen.getByText(/^password must be at least 8 characters/i)).toBeInTheDocument();
    });

    it("login is called and user is redirected to / on submit, when all inputs are valid", async () => {
        const mockNavigate = vi.fn();
        useNavigate.mockReturnValue(mockNavigate);

        const EMAIL = "test@gmail.com";
        const PASSWORD = "12345678";

        renderComponent();

        await userEvent.type(screen.getByLabelText(/^email$/i), EMAIL);
        await userEvent.type(screen.getByLabelText(/^password$/i), PASSWORD);

        await userEvent.click(screen.getByRole("button"));

        expect(login).toHaveBeenCalled();
        expect(login).toHaveBeenCalledWith({ "email": EMAIL, "password": PASSWORD });

        expect(mockNavigate).toHaveBeenCalled();
        expect(mockNavigate).toHaveBeenCalledWith("/");
    });

    it("login is called and user is redirected to redirect_url if present in search params on submit, when all inputs are valid", async () => {
        const EMAIL = "test@gmail.com";
        const PASSWORD = "12345678";

        renderComponent("/chat/123");

        await userEvent.type(screen.getByLabelText(/^email$/i), EMAIL);
        await userEvent.type(screen.getByLabelText(/^password$/i), PASSWORD);

        await userEvent.click(screen.getByRole("button"));

        expect(login).toHaveBeenCalled();
        expect(login).toHaveBeenCalledWith({ "email": EMAIL, "password": PASSWORD })
    });

    it("login is called and error message is shown if invalid credentials are provided", async () => {
        const mockNavigate = vi.fn();
        useNavigate.mockReturnValue(mockNavigate);

        login.mockReturnValue({});

        const EMAIL = "test@gmail.com";
        const PASSWORD = "12345678";

        renderComponent();

        await userEvent.type(screen.getByLabelText(/^email$/i), EMAIL);
        await userEvent.type(screen.getByLabelText(/^password$/i), PASSWORD);

        await userEvent.click(screen.getByRole("button"));

        expect(login).toHaveBeenCalled();
        expect(login).toHaveBeenCalledWith({ "email": EMAIL, "password": PASSWORD });

        expect(screen.getByText(/^invalid credentials$/i)).toBeInTheDocument();
    });
})