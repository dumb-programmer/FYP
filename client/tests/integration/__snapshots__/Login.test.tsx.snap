// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Login > renders correctly 1`] = `
<div>
  <main
    class="min-h-screen w-screen flex justify-center items-center"
  >
    <div
      class="card p-10 shadow-md"
    >
      <form
        class="flex flex-col gap-4 md:min-w-96"
      >
        <h1
          class="card-title text-2xl"
        >
          Login
        </h1>
        <div
          class="form-control"
        >
          <label
            class="label"
            for="email"
          >
            Email
          </label>
          <input
            class="input input-bordered"
            id="email"
            name="email"
          />
          <span
            class="text-sm text-red-600"
          />
        </div>
        <div
          class="form-control"
        >
          <label
            class="label"
            for="password"
          >
            Password
          </label>
          <input
            class="input input-bordered"
            id="password"
            name="password"
            type="password"
          />
          <span
            class="text-sm text-red-600"
          />
        </div>
        <span
          class="text-sm text-red-600"
        />
        <button
          class="btn btn-primary"
        >
          Submit
        </button>
        <p
          class="text-gray-500"
        >
          Don't have an account? 
          <a
            class="link"
            href="/signup"
          >
            Signup
          </a>
        </p>
        <div
          class="divider"
        >
          OR
        </div>
        <div
          class="flex flex-col gap-4"
          data-testid="auth-provider-btns"
        >
          <a
            class="btn glass flex items-center gap-4"
            href="http://localhost:3000/auth/google"
          >
            <img
              alt="google"
              src="/src/assets/google-icon.svg"
              style="transform: scale(1.3);"
            />
            Continue With Google
          </a>
          <a
            class="btn glass flex items-center gap-4"
            href="http://localhost:3000/auth/github"
          >
            <img
              alt="github"
              src="/src/assets/github-icon.svg"
              style="transform: scale(1.8);"
            />
            Continue With Github
          </a>
        </div>
      </form>
    </div>
  </main>
</div>
`;
